src/panels/lovelace/views/hui-view.ts

```diff
@@ -226,59 +226,64 @@ export class HUIView extends LitElement {
       root.removeChild(root.lastChild);
     }
 
-    let columns: [number, number][][] = [];
     const columnEntityCount: number[] = [];
+    const columnElements: HTMLDivElement[] = [];
     for (let i = 0; i < this.columns!; i++) {
-      columns.push([]);
+      const columnEl = document.createElement("div");
+      columnEl.classList.add("column");
+      root.appendChild(columnEl);
       columnEntityCount.push(0);
+      columnElements.push(columnEl);
     }
 
-    this._cards.forEach((el, index) => {
-      const cardSize = computeCardSize(
+    let cardsAdded = 0;
+
+    for (const [index, el] of this._cards.entries()) {
+      const cardSizeProm = computeCardSize(
         (el.tagName === "HUI-CARD-OPTIONS" ? el.firstChild : el) as LovelaceCard
       );
-      columns[getColumnIndex(columnEntityCount, cardSize)].push([
+      let waitProm: Promise<unknown>;
+      if (cardsAdded > 5) {
```

Comment:
So in the old approach each column had their own counter and it would pause for a render after it had added more than 5 "card size units". 

Now we're doing 5 cards, which could be a lot when we're doing vertical stack cards etc. 

I wonder if we can do a new approach here where we just check the time elapsed and if it's time for a pause?

```ts
let start = new Date()
const fps = 60;
const timePerFrame = 1000 / fps;
if (new Date() - start > timePerFrame) {
  waitProm = …
}
``` 

Awaiting card sizes can take until after next render, in that case we are not resetting `cardsAdded`. I wonder if we should treat that differently.

Untested, but would something like this work:

```ts
let tillNextRender = undefined;
let start = undefined;

for (const [index, el] of this._cards.entries()) {
  if (tillNextRender === undefined) {
    tillNextRender = nextRender().then(() => {
      tillNextRender = undefined;
      start = undefined;
    })
  }
    
  let waitProm = undefined;

  if (start === undefined) {
    start = new Date()
    waitProm = Promise.resolve()
  } else if (new Date() - start > 16) {
    waitProm = tillNextRender
  } else {
    waitProm = Promise.resolve()
  }

  const cardSizeProm = computeCardSize(…);
  const [cardSize] = await Promise.all([cardSizeProm, waitProm]);
}
```

Reply:
I like it, takes the guessing out of how many cards we are rendering and how expensive they are.